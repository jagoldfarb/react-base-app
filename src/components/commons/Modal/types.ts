import { ReactNode } from 'react';

export interface IModalProps {
  children: ReactNode;
  disabled?: boolean;
  isOpen: boolean;
  onClose?: () => void;
}
