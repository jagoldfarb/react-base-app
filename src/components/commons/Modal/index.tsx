import React, { FunctionComponent, useEffect } from 'react';
import ReactModal from 'react-modal';

import { COLORS } from 'helpers/enums/colors';

import { IModalProps } from './types';
import {
  ModalContainer,
  ModalContent,
  CloseModalContainer,
  CloseModalIcon,
  ChildrenContainer,
} from './styles';

ReactModal.defaultStyles.overlay = {
  ...ReactModal.defaultStyles.overlay,
  backgroundColor: COLORS.modalBackground,
  zIndex: 10,
};

const Modal: FunctionComponent<IModalProps> = (props: IModalProps) => {
  const { children, disabled = false, isOpen, onClose = () => {} } = props;

  useEffect(() => {
    // 27: Escape key
    const handleKeyUp = (event: KeyboardEvent) => event.keyCode === 27 && !disabled && onClose();
    window.addEventListener('keyup', handleKeyUp, false);
    return () => {
      window.removeEventListener('keyup', handleKeyUp, false);
    };
  });

  return (
    <ModalContainer ariaHideApp={false} isOpen={isOpen}>
      <ModalContent data-testid="modal">
        {!disabled && (
          <CloseModalContainer>
            <CloseModalIcon data-testid="modal-close-button" onClick={onClose} />
          </CloseModalContainer>
        )}
        <ChildrenContainer>{children}</ChildrenContainer>
      </ModalContent>
    </ModalContainer>
  );
};

export default Modal;
