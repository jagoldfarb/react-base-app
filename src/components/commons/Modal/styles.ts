import styled from 'styled-components';
import Modal from 'react-modal';

import { SIZES } from 'helpers/enums/sizes';
import { COLORS } from 'helpers/enums/colors';

import { ReactComponent as CloseModalSVG } from './assets/close-modal.svg';

export const ModalContainer = styled(Modal)`
  && {
    display: flex;
    height: 100%;
    align-items: center;
    justify-content: center;
    text-align: center;
    margin: 0 auto;
    outline: none;

    @media (max-width: ${SIZES.bigPhone}px) {
      max-width: 100%;
      margin: 0 20px;
    }
  }
`;

export const ModalContent = styled.div`
  background-color: ${COLORS.white};
  padding: 10px;
  border-radius: 6px;
  max-height: 90vh;
  overflow: scroll;

  &::-webkit-scrollbar {
    display: none;
  }

  & {
    -ms-overflow-style: none; /* IE and Edge */
    scrollbar-width: none; /* Firefox */
  }
`;

export const CloseModalContainer = styled.div`
  display: flex;
  justify-content: flex-end;
  width: 100%;
`;

export const CloseModalIcon = styled(CloseModalSVG)`
  cursor: pointer;
`;

export const ChildrenContainer = styled.div`
  padding: 30px;

  @media (max-width: ${SIZES.mediumPhone}px) {
    padding: 15px;
  }
`;
