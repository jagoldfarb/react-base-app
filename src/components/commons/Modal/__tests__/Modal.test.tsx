import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Modal from '..';

const defaultProps = {
  children: <span>I am a modal</span>,
  isOpen: true,
  onClose: jest.fn(),
};

describe('Modal Component', () => {
  it('Should render the modal opened', () => {
    const { asFragment, getByTestId, getByText } = render(<Modal {...defaultProps} />);
    expect(getByTestId('modal')).toBeDefined();
    expect(getByTestId('modal-close-button')).toBeDefined();
    expect(getByText('I am a modal')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the modal opened and click close', () => {
    const { asFragment, getByTestId } = render(<Modal {...defaultProps} />);
    expect(getByTestId('modal')).toBeDefined();
    const closeButton = getByTestId('modal-close-button');
    expect(closeButton).toBeDefined();
    fireEvent.click(closeButton);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the modal and press escape', () => {
    const { asFragment, getByTestId } = render(<Modal {...defaultProps} />);
    const modal = getByTestId('modal');
    expect(modal).toBeDefined();
    fireEvent.keyUp(modal, { key: 'Escape', code: 'Escape', keyCode: 27 });
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the modal and press enter', () => {
    const { asFragment, getByTestId } = render(<Modal {...defaultProps} />);
    const modal = getByTestId('modal');
    expect(modal).toBeDefined();
    fireEvent.keyUp(modal, { key: 'Enter', code: 'Enter', keyCode: 13 });
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the modal disabled', () => {
    const props = { ...defaultProps, disabled: true, onClose: undefined };
    const { asFragment, getByTestId, getByText, queryByTestId } = render(<Modal {...props} />);
    expect(getByTestId('modal')).toBeDefined();
    expect(queryByTestId('modal-close-button')).toBeNull();
    expect(getByText('I am a modal')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render the modal opened and click close without functionality', () => {
    const props = { ...defaultProps, onClose: undefined };
    const { asFragment, getByTestId } = render(<Modal {...props} />);
    expect(getByTestId('modal')).toBeDefined();
    const closeButton = getByTestId('modal-close-button');
    expect(closeButton).toBeDefined();
    fireEvent.click(closeButton);
    expect(asFragment()).toMatchSnapshot();
  });
});
