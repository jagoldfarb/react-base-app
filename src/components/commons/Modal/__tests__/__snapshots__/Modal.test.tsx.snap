// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component Should render the modal and press enter 1`] = `<DocumentFragment />`;

exports[`Modal Component Should render the modal and press escape 1`] = `<DocumentFragment />`;

exports[`Modal Component Should render the modal disabled 1`] = `<DocumentFragment />`;

exports[`Modal Component Should render the modal opened 1`] = `<DocumentFragment />`;

exports[`Modal Component Should render the modal opened and click close 1`] = `<DocumentFragment />`;

exports[`Modal Component Should render the modal opened and click close without functionality 1`] = `<DocumentFragment />`;
